### ISSUE TITLE
Different order of items in `Union[Callable[[], T], Callable[[], Awaitable[T]]]` results in different behavior

### ISSUE THREAD

**Issue by wch** [microsoft/pyright#3813]
This is kind of a tricky issue that showed up with Pyright 1.1.266. I can't say I fully understand it, but I'll illustrate with an example.

I have a function that takes a parameter with type:

```
Union[Callable[[], T], Callable[[], Awaitable[T]]]
```

When the order of the types in the `Union` is reversed, like this:

```
Union[Callable[[], Awaitable[T]], Callable[[], T]])
```

... then Pyright has different behavior with type narrowing. This started happening with 1.1.266.

Here's an example that shows the change in behavior. The function `call2` previously did not result in Pyright errors, but it does now.

```py
# pyright: reportUnusedExpression=false

from typing import Callable, Union, Awaitable, TypeVar, cast
from typing_extensions import TypeGuard
import inspect

T = TypeVar("T")


def is_async_callable(
    obj: Union[Callable[..., T], Callable[..., Awaitable[T]]]
) -> TypeGuard[Callable[..., Awaitable[T]]]:
    """
    Returns True if `obj` is an `async def` function, or if it's an object with a
    `__call__` method which is an `async def` function.
    """
    if inspect.iscoroutinefunction(obj):
        return True
    if hasattr(obj, "__call__"):  # noqa: B004
        if inspect.iscoroutinefunction(obj.__call__):  # type: ignore
            return True

    return False


async def call(fn: Union[Callable[[], T], Callable[[], Awaitable[T]]]) -> T:
    if is_async_callable(fn):
        fn
        # Inferred type with 1.1.265: (parameter) fn: (...) -> Awaitable[T@call]
        # Inferred type with 1.1.266: (parameter) fn: (...) -> Awaitable[T@call]

        return await fn()  # OK
    else:
        # cast() is needed because TypeGuard doesn't support negative type narrowing.
        # (This is expected behavior)
        fn = cast(Callable[[], T], fn)
        return fn()


async def call2(fn: Union[Callable[[], Awaitable[T]], Callable[[], T]]) -> T:
    if is_async_callable(fn):
        fn
        # Inferred type with 1.1.265: (parameter) fn: (...) -> Awaitable[T@call2]
        # Inferred type with 1.1.266: (parameter) fn: (...) -> Awaitable[Awaitable[T@call2] | T@call2]

        return await fn()
        #      ^^^^^^^^^^
        # Expression of type "Awaitable[T@call2] | T@call2" cannot be assigned to return type "T@call2"
        # Type "Awaitable[T@call2] | T@call2" cannot be assigned to type "T@call2"
    else:
        fn = cast(Callable[[], T], fn)
        return fn()
```

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is fixed in pyright 1.1.267, which I just published. It will also be included in a future version of pylance.
